src/Hooks/useAsyncFetch.ts

```diff
@@ -0,0 +1,101 @@
+import { useEffect, useReducer, useRef, useCallback } from "react";
+
+const START_FETCH = "START_FETCH";
+const FETCH_SUCCESS = "FETCH_SUCCESS";
+const FETCH_FAILED = "FETCH_FAILED";
+
+interface ResultObject<T> {
+  value: T;
+  error: any;
+  isPending: boolean;
+}
+
+type Action<T> =
+  | {
+      type: "START_FETCH";
+    }
+  | {
+      type: "FETCH_SUCCESS";
+      value: T;
+    }
+  | {
+      type: "FETCH_FAILED";
+      error: any;
+    };
+
+type ReducerFunction<T> = (
+  state: ResultObject<T>,
+  action: Action<T>
+) => ResultObject<T>;
+
+function asyncFetchReducer<T>(
+  state: ResultObject<T>,
+  action: Action<T>
+): ResultObject<T> {
+  switch (action.type) {
+    case START_FETCH:
+      return { ...state, isPending: true };
+    case FETCH_SUCCESS:
+      return { ...state, value: action.value, isPending: false };
+    case FETCH_FAILED:
+      return { ...state, error: action.error, isPending: false };
+    default:
+      return state;
+  }
+}
+
+export default function useAsyncFetch<ResultType>(
+  queryFunc: (params?: any) => Promise<ResultType>,
+  params?: object | null,
+  defaultValue?: ResultType
```

Comment:
Could you explain what defaultValue is/does? Having a hard time following along :P

Reply:
It's what value is initially set to, similar to useState(defaultValue)